### DISCUSSION TITLE
Is it possible to control the onboard LED and button?

### DISCUSSION THREAD

**Discussion by chaotix-** [Peter-van-Tol/LiteX-CNC#116] [Q&A] [answered]
I was wondering if it was possible to add the onboard LED and possibly the button on the Colorlight boards to a GPIO instance?

In https://github.com/litex-hub/litex-boards/blob/master/litex_boards/platforms/colorlight_5a_75e.py I found that the IO pins are called "user_led_n" and "user_btn_n". But using this in LiteX-CNC in a pin definition instead of "j1:1" throws an error when building the firmware. I also tried "user_led:0" and "user_led0".

I thought it would be nice to display a heartbeat with the onboard LED to show that the connection to LinuxCNC is working.

**Comment by Peter-van-Tol** [comment]
I like the idea. This should be possible with the 5A-75X boards, if you use the HUB75HAT i'm not sure whether the pjns are exported.

Next week I'll try it acces those pins on my 5A-75B and fix it if it is not working (I think I used it as heartbeat before). In any case, I will add some documentation on using these built-in functions.

**Reply by Peter-van-Tol** [reply]
As an addition, i can add some functionality to the watchdog to output a heartbeat with configurable frequency and dutycycle (then you don't need to abuse a PWM for that).

When implemented, are you willing to test?

**Reply by Peter-van-Tol** [reply]
The user LED and button cannot be controlled at this moment. Reason is that these are not placed on connectors, but are dedicated io (i.e. in Litex there is already a module which grabbed this pin). Connectors and io cannot be mixed, which is a bummer.

As a solution I will modify the board definition used by LiteX-CNC. I will move the led and button to a connector and then they can be used.

A check will be added that the LED can only be used for output and that the button is limited to input. This to prevent damage to the board when it is misconfigured.

**Reply by Peter-van-Tol** [reply]
I created a new branch to add support for user LED and button: #117 

I added now support for the user LED and button on 5a-75b, 5a-75e and HUB75HAT boards. Remaining issues are:
- the SPI was using the user LED for debug. It is now commented out, but I want to have the debug support back;
- Implement restrictions:
  - the LED will be supported in GPIO - out and PWM only. 
  - the button will be supported in GPIO - in only.

The restrictions are required to prevent damage when the card is misconfigured. The button can short the card if by accident it is set as output.

An example is included in the issue.

**Reply by Peter-van-Tol** [reply]
#117 has been merged into `main`.

**Comment by chaotix-** [comment]
I would be willing to test the watchdog solution. But I think the signal should be generated by LinuxCNC to show that the connection is working. When the FPGA outputs the signal it just shows that the FPGA is doing something.
Yes it should stop blinking after the connection to LinuxCNC broke and the watchdog tripped but if there is a problem with tripping the watchdog the LED will blink anyway although there is no connection to LinuxCNC.

I tried to generate a heartbeat signal with HAL, but ran into some problems when using pwmgen:
```
loadrt pwmgen output_type=0

addf pwmgen.update servo-thread
addf pwmgen.make-pulses servo-thread

setp pwmgen.0.value 0.5
setp pwmgen.0.enable 1
setp pwmgen.0.pwm-freq 0.5
net heatbeat_square [LITEXCNC](NAME).gpio.output1.out <= pwmgen.0.pwm
```
`setp pwmgen.0.pwm-freq 0.5` does not throw an error but it also does not set the frequency. It's still '0' afterwards and the PWM-frequency is much to high so an LED connected to the output looks dimmmed but is not blinking (for the human eye). When using halshow to set pwm-freq to 0.5 by hand it works just fine. I have no idea why.

Using siggen to generate a sine wave to get a nice pulsing works just fine:
```
loadrt siggen num_chan=1

addf siggen.0.update servo-thread

setp siggen.0.amplitude 0.5
setp siggen.0.offset 0.5
setp siggen.0.frequency 0.5
setp [LITEXCNC](NAME).pwm.00.pwm_freq 100
setp [LITEXCNC](NAME).pwm.00.enable 1
net heatbeat_sine [LITEXCNC](NAME).pwm.00.value <= siggen.0.sine
```

Maybe this is useful to someone.

**Reply by Peter-van-Tol** [reply]
The pwm frequency is defined as an integer. Setting it a value of 0.5, LinuxCNC will implicitly round it down to 0. A value of 0 will enable PDM for that channel, which indeed has a very high frequency. 

Technically speaking, the minimum PWM frequencybis 1 Hz.

The challenge of using PWM as heartbeat for communication, is that when communication is lost, the PWM cannot be disabled.the new state can most likely not be send to the FPGA in that case.

Therefor, integrating the heart beat in the watchdog, will stop the heartbeat as soon the dogs bites due to messages which have not been received. 

The watchdog already has physical pin which is low when watchdog is not happy. I use this pin to turn the buffers off, physically making it impossible to send signals to the machine.

**Reply by chaotix-** [reply]
> The pwm frequency is defined as an integer. Setting it a value of 0.5, LinuxCNC will implicitly round it down to 0. A value of 0 will enable PDM for that channel, which indeed has a very high frequency.
> 
> Technically speaking, the minimum PWM frequencybis 1 Hz.

This is not correct. `pwmgen.N.pwm-freq` is documented as a float. I also tested it with `setp pwmgen.0.pwm-freq 1` which does not make a difference. Furthermore setting it to 0.5 with halshow works just fine.

> The challenge of using PWM as heartbeat for communication, is that when communication is lost, the PWM cannot be disabled.the new state can most likely not be send to the FPGA in that case.

This is exactly the reason why I think it would be advantageous to generate the PWM signal with LinuxCNC for example with siggen. As soon as the communication drops the LED will freeze. (You are right that using a PWM instance on the FPGA to generate the change in brightness obviously makes no sense. In the example the PWM instance is used to set one specific brightness that is changed by LinuxCNC with every update to the servo-thread. Upon losing the connection the LED will be stuck with the last brightness set by LinuxCNC and not pulsate anymore.)

> Therefor, integrating the heart beat in the watchdog, will stop the heartbeat as soon the dogs bites due to messages which have not been received.
> 
> The watchdog already has physical pin which is low when watchdog is not happy. I use this pin to turn the buffers off, physically making it impossible to send signals to the machine.

Yes. As long as the watchdog works as intended this will be just fine. In practice it will probably not make a huge difference. Tying the heartbeat to the watchdog will show if the watchdog is not triggered. Tying the heartbeat to a HAL pin will show if the communication with LinuxCNC is working. This should in fact be the same. But if something goes wrong with the watchdog it might not be. As improbable as this scenario might be.

One more difference might be in the startup. I have not tested the watchdog yet. So I don't know if this is the case. If the watchdog is happy before establishing a connection to LinuxCNC it might show a heartbeat although communication has not been established yet. But this in pure speculation and also depends on how you would implement that feature.

**Reply by Peter-van-Tol** [reply]
I have tonsrand corrected.

The frequency is indeed a float. However, its lower bound is set to 1 Hz and there is no message when this happens.

I will decrease the minimum frequency to 0.1 Herz (10 seconds) and will add a message when this value is modified due to bounds. The message will be shown once (or can be reset using a HAL pin).

A frequency of 1 HZ should be visible by the human eye, o have to check why it was dim.

**Reply by chaotix-** [reply]
I think we are talking about different things. I wrote about a problem with the `pwmgen.N.pwm-freq` HAL pin from the LinuxCNC HAL component `pwmgen` as documented here: https://linuxcnc.org/docs/html/man/man9/pwmgen.9.html
I guess you are writing about the PWM generator that is running on the FPGA.
Sorry about the confusion.

**Reply by Peter-van-Tol** [reply]
No worries. It inspired me to rewrite the `watchdog` module. For the HUB75HAT I require a enable signal, but I also want to have a heartbeat. I'm adding function to `watchdog`, so one can add functions when the watchdog is happy or has bitten. See also #119

**Comment by chaotix-** [comment, marked as answer]
LED and button now work in `main` branch.
